pkg/client/compose/port.go

```diff
@@ -41,3 +76,119 @@ func transformPortsExtension(ports PortsSource) ([]api.PortSpec, error) {
 
 	return specs, nil
 }
+
+// expandPortRange expands a port range like "3000-3005" into individual port configs
+func expandPortRange(basePort types.ServicePortConfig) ([]types.ServicePortConfig, error) {
+	if !strings.Contains(basePort.Published, "-") {
+		return []types.ServicePortConfig{basePort}, nil
+	}
+
+	parts := strings.Split(basePort.Published, "-")
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid port range format %q", basePort.Published)
+	}
+
+	startPort, err := strconv.ParseUint(strings.TrimSpace(parts[0]), 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("invalid start port in range %q: %w", basePort.Published, err)
+	}
+
+	endPort, err := strconv.ParseUint(strings.TrimSpace(parts[1]), 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("invalid end port in range %q: %w", basePort.Published, err)
+	}
+
+	if startPort >= endPort {
+		return nil, fmt.Errorf("invalid port range %q: start port must be less than end port", basePort.Published)
+	}
+
+	// Expand range into individual ports
+	var expandedPorts []types.ServicePortConfig
+	portDiff := endPort - startPort
+
+	for i := uint64(0); i <= portDiff; i++ {
+		expandedPort := basePort // Copy the base configuration
+
+		// Set published port
+		publishedPort := startPort + i
+		expandedPort.Published = strconv.FormatUint(publishedPort, 10)
+
+		// Set target port (aka container port)
+		// If target port range matches published port range, expand it too
+		// Otherwise, keep the same target port for all
+		if basePort.Target != 0 {
+			// For ranges, Docker Compose maps ports 1:1 by default
+			expandedPort.Target = basePort.Target + uint32(i)
+		}
+
+		expandedPorts = append(expandedPorts, expandedPort)
+	}
+
+	return expandedPorts, nil
+}
+
+// convertServicePortConfigToPortSpec converts a single ServicePortConfig directly to api.PortSpec
+func convertServicePortConfigToPortSpec(port types.ServicePortConfig) (api.PortSpec, error) {
+	spec := api.PortSpec{
+		ContainerPort: uint16(port.Target),
+		Protocol:      port.Protocol,
+		Mode:          port.Mode,
+	}
+
+	// Apply defaults according to Docker Compose specification
+	if spec.Protocol == "" {
+		spec.Protocol = "tcp"
+	}
+	if spec.Mode == "" {
+		spec.Mode = "ingress"
```

Comment:
Oh, I just realised that the meaning of the `host` [mode](https://github.com/compose-spec/compose-spec/blob/main/05-services.md#long-syntax-3) in Compose spec differs from the ones in Uncloud. Uncloud publishes the host port only on the machines that the service runs on, *not* on all the machines.

Also, the compose default mode is [ingress](https://github.com/compose-spec/compose-go/blob/main/types/types.go#L531) and `ports` only supports `tcp` and `udp` ports. But Uncloud supports ingress (expose via the reverse proxy) only for the `http(s)` ports at the moment. This means that any `ports` using short syntax will result in unsupported service spec for uncloud which undermines the entire effort to support `ports` spec (at least for now) 🤔